core/src/sumeragi/main_loop.rs

```diff
@@ -605,16 +593,34 @@ fn handle_message<F: FaultInjection>(
         (Message::ViewChangeSuggested, _) => {
             trace!("Received view change suggestion.");
         }
-        (Message::BlockSyncUpdate(BlockSyncUpdate { block }), _) => {
+        (Message::BlockSyncUpdate(block_sync_update), _) => {
+            let block = block_sync_update.extract_block_unchecked();
             info!(%addr, %role, hash=%block.hash(), "Block sync update received");
 
-            let block = match block.revalidate(current_topology) {
-                Ok(block) => block,
-                Err((block, error)) => {
-                    error!(%addr, %role, hash=%block.hash(), ?error, "Block not valid");
+            let block_hash = block.hash();
+            if block
+                .revalidate::<true>(
+                    &sumeragi.transaction_validator,
+                    &state.wsv,
+                    state.latest_block_hash,
+                    state.latest_block_height,
+                )
+                .is_err()
+            {
+                /* If the block fails validation we must check again using the finaziled wsv.
+                When a soft-fork occurs the consensus-block may be valid on the previous
+                wsv but not the current one. */
+                if let Err(error) = block.revalidate::<true>(
+                    &sumeragi.transaction_validator,
+                    &state.finalized_wsv,
+                    state.latest_block_hash,
+                    state.latest_block_height,
```

Comment:
If we are revalidating against `finalized_wsv` shouldn't we use appropriate `block_height` and `block_hash`?

Reply:
true!